internal/deployment_client.go

```diff
@@ -0,0 +1,243 @@
+		// service/application are linked together by sharing a series name.
+		SeriesName string
+
+		// BuildID - identifies the worker's code and configuration version.
+		BuildID string
+	}
+
+	// DeploymentTaskQueueInfo describes properties of the Task Queues involved
+	// in a deployment.
+	// NOTE: Experimental
+	DeploymentTaskQueueInfo struct {
+		// Name - Task queue name.
+		Name string
+
+		// Type - The type of this task queue.
+		Type TaskQueueType
+
+		// FirstPollerTime - Time when the server saw the first poller for this task queue
+		// in this deployment.
+		FirstPollerTime time.Time
+	}
+
+	// DeploymentInfo holds information associated with
+	// workers in this deployment.
+	// Workers can poll multiple task queues in a single deployment,
+	// which are listed in this message.
+	// NOTE: Experimental
+	DeploymentInfo struct {
+		// Deployment - An identifier for this deployment.
+		Deployment Deployment
+
+		// CreateTime - When this deployment was created.
+		CreateTime time.Time
+
+		// Current - Whether this deployment is the current one for its deployment series.
+		Current bool
+
+		// TaskQueuesInfo - List of task queues polled by workers in this deployment.
+		TaskQueuesInfo []DeploymentTaskQueueInfo
+
+		// Metadata - A user-defined set of key-values. Can be updated with `DeploymentClient.SetCurrent`.
+		Metadata map[string]*commonpb.Payload
+	}
+
+	// DeploymentListEntry is a subset of fields from DeploymentInfo
+	// NOTE: Experimental
+	DeploymentListEntry struct {
+		// An identifier for this deployment.
+		Deployment Deployment
+
+		// When this deployment was created.
+		CreateTime time.Time
+
+		// Whether this deployment is the current one for its deployment series.
+		Current bool
+	}
+
+	// DeploymentListIterator is an iterator for deployments.
+	// NOTE: Experimental
+	DeploymentListIterator interface {
+		// HasNext - Return whether this iterator has next value.
+		HasNext() bool
+
+		// Next - Returns the next deployment and error
+		Next() (*DeploymentListEntry, error)
+	}
+
+	// DeploymentListOptions are the parameters for configuring listing deployments.
+	// NOTE: Experimental
+	DeploymentListOptions struct {
+		// PageSize - How many results to fetch from the Server at a time.
+		// Optional: defaulted to 1000
+		PageSize int
+
+		// SeriesName - Filter with the name of the deployment series.
+		// Optional: If present, use an exact series name match.
+		SeriesName string
+	}
+
+	// DeploymentReachabilityInfo extends DeploymentInfo with reachability information.
+	// NOTE: Experimental
+	DeploymentReachabilityInfo struct {
+		DeploymentInfo
+
+		// Reachability - Kind of tasks that may reach a worker
+		// associated with a deployment.
+		Reachability DeploymentReachability
+
+		// LastUpdateTime - When reachability was last computed. Computing reachability
+		// is an expensive operation, and the server caches results.
+		LastUpdateTime time.Time
```

Comment:
hmm.. if we embed `DeploymentInfo` there would be timestamps such `CreateTime` besides this `LastUpdateTime`, user wouldn't know if the field only refers to reachability status update time. Later more timestamps and other fields will be added to Deployment Info (and maybe to Reachability response) and I worry it'd be confusing for users if we mix them. Should we make deployment info just a nested field instead of an embed?